### PR TITLE
INT-229_INT-231_configure_credentials_with_folders

### PR DESCRIPTION
**Pull request by fernau**
Screencast: https://youtu.be/7U0TgR9-0J0
JIRA: https://issues.sonatype.org/browse/INT-229
JIRA: https://issues.sonatype.org/browse/INT-231

### CODE REVIEW

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 1]
```diff
@@ -87,18 +84,26 @@ class IqClientFactory
     }
   }
 
-  static private Authentication loadCredentials(final URI url, final String credentialsId) {
-    def lookupCredentials = CredentialsProvider.lookupCredentials(
-        StandardUsernamePasswordCredentials,
-        (ItemGroup) Jenkins.getInstance(),
+  static private StandardCredentials findCredentials(final URI url, final String credentialsId, final context) {
```
Comment (whyjustin) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: Credentials is @Nullable on [0] it should be Nullable here.

[0] https://github.com/jenkinsci/nexus-platform-plugin/pull/7/files#diff-269157b33e8fa445ec4f21d9778719b2R50

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 2]
```diff
@@ -87,18 +84,26 @@ class IqClientFactory
...
+    //noinspection GroovyAssignabilityCheck
+    List<StandardCredentials> lookupCredentials = CredentialsProvider.lookupCredentials(
+        StandardCredentials,
+        context,
         ACL.SYSTEM,
         URIRequirementBuilder.fromUri(url.toString()).build())
 
     def credentials = CredentialsMatchers.firstOrNull(lookupCredentials, CredentialsMatchers.withId(credentialsId))
```
Comment (whyjustin) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: If credentialsId is null then we should use the global credentialsId?

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 3]
```diff
@@ -23,36 +24,42 @@ import com.sonatype.nexus.api.iq.internal.InternalIqClientBuilder
...
 import jenkins.model.Jenkins
+import org.slf4j.Logger
 
 class IqClientFactory
 {
-  static InternalIqClient getIqClient() {
-    return getIqClient(NxiqConfiguration.serverUrl, (String) NxiqConfiguration.credentialsId)
+  static InternalIqClient getIqClient(String credentialsId, Item context) {
+    // credentialsId can be an empty String if unset. Empty strings are falsy in Groovy
+    def credentials = credentialsId ? findCredentials(NxiqConfiguration.serverUrl, credentialsId, context)
```
Comment (whyjustin) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: If `static private StandardCredentials findCredentials(final URI url, final String credentialsId, final context)` if refactored to handle Nullable credentials then can it be called here to get credentials?

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 4]
```diff
@@ -87,18 +84,26 @@ class IqClientFactory
...
+        StandardCredentials,
+        context,
         ACL.SYSTEM,
         URIRequirementBuilder.fromUri(url.toString()).build())
 
     def credentials = CredentialsMatchers.firstOrNull(lookupCredentials, CredentialsMatchers.withId(credentialsId))
-    if (!credentials) {
-      throw new IllegalArgumentException(Messages.IqClientFactory_NoCredentials(credentialsId))
-    }
+    Optional.ofNullable(credentials)
```
Comment (whyjustin) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: I'm guessing Optional<StandardCredentials> casts implicitly to StandardCredentials? This seems a very roundabout way of just writing
```
if (!credentials) {
  throw
}
```

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy [hunk 5]
```diff
@@ -46,7 +46,7 @@ class IqPolicyEvaluatorUtil
       loggerBridge.debug(Messages.IqPolicyEvaluation_Evaluating())
 
       def credentialsId = getCredentials(iqPolicyEvaluator.jobCredentialsId)
-      def iqClient = IqClientFactory.getIqClient(loggerBridge, credentialsId)
+      def iqClient = IqClientFactory.getIqClient(loggerBridge, run.parent.parent, credentialsId)
```
Comment (whyjustin) [inline src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: Why parent.parent? I'm wondering because there can be n-tiers in the hierarchy.

File: src/main/resources/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorWorkflowStep/config.groovy [hunk 6]
```diff
@@ -64,16 +64,8 @@ f.section(title: descriptor.displayName) {
         f.checkbox()
       }
 
-      if (!nxiqConfiguration?.@isPkiAuthentication) {
-        f.entry(title: _(Messages.IqPolicyEvaluation_JobSpecificCredentials()), field: 'jobCredentialsId') {
```
Comment (whyjustin) [inline src/main/resources/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorWorkflowStep/config.groovy]: Can this string be removed from the Messages? IqPolicyEvaluation_JobSpecificCredentials

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy (hunk 3 above)
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: I have consolidated all getIqClient into a single method. empty strings and null values are no longer considered a valid value. just because jenkins does that doesn't mean we have to continue this pattern. a000634

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 7]
```diff
@@ -23,36 +24,42 @@ import com.sonatype.nexus.api.iq.internal.InternalIqClientBuilder
...
+import org.slf4j.Logger
 
 class IqClientFactory
 {
-  static InternalIqClient getIqClient() {
-    return getIqClient(NxiqConfiguration.serverUrl, (String) NxiqConfiguration.credentialsId)
+  static InternalIqClient getIqClient(String credentialsId, Item context) {
+    // credentialsId can be an empty String if unset. Empty strings are falsy in Groovy
+    def credentials = credentialsId ? findCredentials(NxiqConfiguration.serverUrl, credentialsId, context)
+      : findCredentials(NxiqConfiguration.serverUrl, NxiqConfiguration.credentialsId, Jenkins.instance)
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: I have consolidated all getIqClient into a single method. empty strings and null values are no longer considered a valid value. just because jenkins does that doesn't mean we have to continue this pattern. a000634

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy (hunk 1 above)
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: I have consolidated all getIqClient into a single method. empty strings and null values are no longer considered a valid value. just because jenkins does that doesn't mean we have to continue this pattern. a000634

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy (hunk 2 above)
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: I have consolidated all getIqClient into a single method. empty strings and null values are no longer considered a valid value. just because jenkins does that doesn't mean we have to continue this pattern. a000634

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 5 above)
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: Replaced with run.parent which is always JobT.
long story: initially I used Project as context but when implementing authn for pipeline switched to Job because this works for both pipeline as well as freestyle job. However here I've forgotten to update it.

File: src/main/resources/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorWorkflowStep/config.groovy (hunk 6 above)
Comment (fernau) [inline src/main/resources/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorWorkflowStep/config.groovy]: removed both unused pki related messages in 622587c

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 8]
```diff
@@ -25,38 +24,29 @@ import org.sonatype.nexus.ci.util.ProxyUtil
...
-    return (InternalIqClient) InternalIqClientBuilder.create()
-        .withServerConfig(getServerConfig(serverUrl, credentialsId))
-        .withProxyConfig(getProxyConfig(serverUrl))
-        .build()
-  }
-
-  static InternalIqClient getIqClient(Logger log, @Nullable String credentialsId) {
-    def serverConfig = getServerConfig(NxiqConfiguration.serverUrl, credentialsId ?: NxiqConfiguration.credentialsId)
-    def proxyConfig = getProxyConfig(NxiqConfiguration.serverUrl)
+  static InternalIqClient getIqClient(Map conf, String credentialsId) {
```
Comment (whyjustin) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: I would vote for a POJO over Map. Dynamic types are powerful but very prone to fat finger mistakes. W/O compile time hints and autocomplete (a la C# dynamics) I am more comfortable with strong types.

Comment (whyjustin) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: I would vote for a POJO over Map. Dynamic types are powerful but very prone to fat finger mistakes. W/O compile time hints and autocomplete (a la C# dynamics) I am more comfortable with strong types.

**Comment by bigspotteddog** [comment]
Codenarc seems to be failing on itself? (does not fail on master):

```
Tests run: 5, Failures: 0, Errors: 5, Skipped: 0, Time elapsed: 12.854 sec <<< FAILURE! - in org.sonatype.nexus.ci.quality.CodeQualityTest
perform static analysis of groovy files using CodeNarc(org.sonatype.nexus.ci.quality.CodeQualityTest)  Time elapsed: 5.577 sec  <<< ERROR!
: Exceeded maximum number of priority 2 violations: (p1=0; p2=8; p3=2)
```

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy (hunk 8 above)
Comment (bigspotteddog) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: I vote for POJO as well.

Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: This is not a map but groovy named parameters which are converted to a map. intellij does have autocomplete for groovy named parameters. It actually analyses the source code and gives you a list of parameters names. however it is not strongly typed. replaced with boring POJO.

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy (hunk 4 above)
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: replaced with roundabout way of checking credentials with Preconditions.checkArgument()

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 9]
```diff
@@ -87,18 +68,32 @@ class IqClientFactory
...
   }
 
-  static private Authentication loadCredentials(final URI url, final String credentialsId) {
-    def lookupCredentials = CredentialsProvider.lookupCredentials(
-        StandardUsernamePasswordCredentials,
-        (ItemGroup) Jenkins.getInstance(),
+  static private StandardCredentials findCredentials(final URI url, final String credentialsId, final context) {
+    checkNotNull(credentialsId)
+    checkNotNull(context)
+    checkArgument(!credentialsId.isEmpty())
```
Comment (whyjustin) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: This is not the same behavior as before. Before checked in credentials was null, this checks if credentialsId is empty. `CredentialsMatchers.firstOrNull` could return null even if credentialsId has a value.

Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: I'm not sure i understand.
This checks credentialsId for null as well as empty string
When CredentialsMatchers.firstOrNull returns null it will throw an InvalidArgumentException

File: src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties [hunk 10]
```diff
@@ -23,8 +24,6 @@ IqPolicyEvaluation.Application=Application
 IqPolicyEvaluation.AdvancedOptions=Advanced options
 IqPolicyEvaluation.ScanPatterns=Scan files in workspace. Wildcards allowed e.g: **/*.jar
 IqPolicyEvaluation.FailOnNetwork=Fail build when unable to communicate with IQ Server
-IqPolicyEvaluation.JobSpecificCredentials=Use job specific credentials
```
Comment (bigspotteddog) [inline src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: Looks like we are using this one and it cannot be removed.

Comment (fernau) [inline src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: restored previously deleted message 1b96583

File: src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy [hunk 11]
```diff
@@ -63,33 +66,16 @@ class IqClientFactoryTest
...
       CredentialsMatchers.firstOrNull(_, _) >> credentials
     when:
-      IqClient client = IqClientFactory.getIqClient(URI.create(url), credentialsId)
+      IqClient client = IqClientFactory.getIqClient(
+          new IqClientFactoryConf(credentialsId: credentialsId, serverUrl: URI.create(url)))
     then:
       client != null
   }
 
-  def 'it uses configured serverUrl and credentialsId'() {
```
Comment (whyjustin) [inline src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy]: What happened to this test?

File: src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy [hunk 12]
```diff
@@ -154,7 +144,7 @@ class IqClientFactoryTest
       InternalIqClientBuilder.create() >> iqClientBuilder
 
     when:
-      IqClientFactory.getIqClient((Logger)null, 'job-specific-creds')
+      IqClientFactory.getIqClient(new IqClientFactoryConf(credentialsId: 'job-specific-creds'))
```
Comment (whyjustin) [inline src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy]: Maybe not part of this PR but this doesn't seem to fit the test name 'it uses logger and job specific credentialsId when provided'

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 13]
```diff
@@ -87,18 +68,32 @@ class IqClientFactory
     }
   }
 
-  static private Authentication loadCredentials(final URI url, final String credentialsId) {
-    def lookupCredentials = CredentialsProvider.lookupCredentials(
-        StandardUsernamePasswordCredentials,
-        (ItemGroup) Jenkins.getInstance(),
+  static private StandardCredentials findCredentials(final URI url, final String credentialsId, final context) {
+    checkNotNull(credentialsId)
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: Added null check for URL

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 14]
```diff
@@ -87,18 +68,32 @@ class IqClientFactory
...
+    checkNotNull(context)
+    checkArgument(!credentialsId.isEmpty())
+
+    //noinspection GroovyAssignabilityCheck
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: A straight cast isn't enough in this case because this is a dynamic call where groovy works out at runtime which is the correct method to call.
to make the AssignablityCheck go away we would use instanceof and then branch for every possible type (ItemGroup or Item).

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy [hunk 15]
```diff
@@ -46,7 +46,9 @@ class IqPolicyEvaluatorUtil
       loggerBridge.debug(Messages.IqPolicyEvaluation_Evaluating())
 
       def credentialsId = getCredentials(iqPolicyEvaluator.jobCredentialsId)
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: IqPolicyEvaluator.getCredentials seems redundant and is now gone

File: src/main/java/org/sonatype/nexus/ci/util/IqUtil.groovy [hunk 16]
```diff
@@ -16,33 +16,42 @@ import javax.annotation.Nullable
...
+        new IqClientFactoryConf(credentialsId: credentialsId, context: context, serverUrl: URI.create(serverUrl)))
     return client.getApplicationsForApplicationEvaluation()
   }
 
-  static ListBoxModel doFillIqStageItems(@Nullable final String credentialsId) {
+  static ListBoxModel doFillIqStageItems(@Nullable final String credentialsId, final Job job) {
     if (NxiqConfiguration.iqConfig) {
-      def client = IqClientFactory.getIqClient(credentialsId)
+      def client = IqClientFactory.getIqClient(
+          new IqClientFactoryConf(credentialsId: credentialsId ?: NxiqConfiguration.credentialsId, context: job))
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/util/IqUtil.groovy]: redundant fallback to default credentials removed

File: src/main/java/org/sonatype/nexus/ci/util/IqUtil.groovy [hunk 17]
```diff
@@ -16,33 +16,42 @@ import javax.annotation.Nullable
...
       FormUtil.newListBoxModelWithEmptyOption()
     }
   }
 
-  static ListBoxModel doFillIqApplicationItems(@Nullable final String credentialsId) {
+  static ListBoxModel doFillIqApplicationItems(@Nullable final String credentialsId, final Job job) {
     if (NxiqConfiguration.iqConfig) {
-      def client = IqClientFactory.getIqClient(credentialsId)
+      def client = IqClientFactory.getIqClient(
+          new IqClientFactoryConf(credentialsId: credentialsId ?: NxiqConfiguration.credentialsId, context: job))
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/util/IqUtil.groovy]: redundant fallback to default credentials removed

File: src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy (hunk 12 above)
Comment (fernau) [inline src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy]: undeleted deleted test 'it uses configured serverUrl and credentialsId'()

File: src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy (hunk 11 above)
Comment (fernau) [inline src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy]: undeleted deleted test 'it uses configured serverUrl and credentialsId'() 66f4013

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 18]
```diff
@@ -25,38 +24,30 @@ import org.sonatype.nexus.ci.util.ProxyUtil
...
-        .withServerConfig(getServerConfig(serverUrl, credentialsId))
-        .withProxyConfig(getProxyConfig(serverUrl))
-        .build()
-  }
-
-  static InternalIqClient getIqClient(Logger log, @Nullable String credentialsId) {
-    def serverConfig = getServerConfig(NxiqConfiguration.serverUrl, credentialsId ?: NxiqConfiguration.credentialsId)
-    def proxyConfig = getProxyConfig(NxiqConfiguration.serverUrl)
+  static InternalIqClient getIqClient(IqClientFactoryConf conf = new IqClientFactoryConf()) {
+    def serverUrl = conf.serverUrl ? conf.serverUrl : NxiqConfiguration.serverUrl
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: Replaced ternary with ?:

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 19]
```diff
@@ -87,18 +68,33 @@ class IqClientFactory
...
-        StandardUsernamePasswordCredentials,
-        (ItemGroup) Jenkins.getInstance(),
+  static private StandardCredentials findCredentials(final URI url, final String credentialsId, final context) {
+    checkNotNull(credentialsId)
+    checkNotNull(context)
+    checkNotNull(url)
+    checkArgument(!credentialsId.isEmpty())
+
+    //noinspection GroovyAssignabilityCheck
+    List<StandardCredentials> lookupCredentials = CredentialsProvider.lookupCredentials(
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: In this case an explicit type is needed. It seems groovy can't derive the type when there are multiple implementations of lookupCredentials

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy [hunk 20]
```diff
@@ -87,18 +68,33 @@ class IqClientFactory
...
+  }
 
-    return new Authentication(credentials.getUsername(), credentials.getPassword().getPlainText())
+  static private ServerConfig getServerConfig(final URI url, final credentials) {
+    if (credentials in StandardUsernamePasswordCredentials) {
+      return new ServerConfig(url, new Authentication(credentials.getUsername(),
+          credentials.getPassword().getPlainText()))
+    } else if (credentials in StandardCertificateCredentials) {
+      return new ServerConfig(url, new CertificateAuthentication(credentials.getKeyStore(),
+          credentials.password.plainText.toCharArray()))
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactory.groovy]: made more use of groovy properties where used inconsistently

File: src/main/java/org/sonatype/nexus/ci/iq/IqClientFactoryConf.groovy [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package org.sonatype.nexus.ci.iq
+
+import hudson.model.ModelObject
+import org.slf4j.Logger
+
+class IqClientFactoryConf
```
Comment (fernau) [inline src/main/java/org/sonatype/nexus/ci/iq/IqClientFactoryConf.groovy]: replaced abbreviated class name with IqClientFactoryConfiguration

File: src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy [hunk 22]
```diff
@@ -125,22 +174,24 @@ class IqClientFactoryTest
...
     when:
-      IqClientFactory.getIqClient('')
+      IqClientFactory.getIqClient(new IqClientFactoryConf(credentialsId: ''))
+
     then:
       1 * iqClientBuilder.withServerConfig { ServerConfig config ->
         config.address == URI.create('http://localhost/')
       } >> iqClientBuilder
       1 * iqClientBuilder.withProxyConfig(_) >> iqClientBuilder
+    iqClientBuilder.withLogger(_) >> iqClientBuilder
```
Comment (fernau) [inline src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy]: whitespace fix

File: src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy [hunk 23]
```diff
@@ -63,11 +72,45 @@ class IqClientFactoryTest
...
+      cert.keyStore >> Mock(KeyStore)
+      cert.password >> secret
+      CredentialsMatchers.firstOrNull(_, _) >> cert
+
+    when:
+      DefaultIqClient client = (DefaultIqClient) IqClientFactory.getIqClient(
+          new IqClientFactoryConf(credentialsId: credentialsId, serverUrl: URI.create(url)))
+
+    then:
+      client != null
```
Comment (fernau) [inline src/test/java/org/sonatype/nexus/ci/iq/IqClientFactoryTest.groovy]: in test added verification that cert and secret are being used
